[Mission5/우창욱] Project_Notion_Vanilla JS 과제

wukdddang opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->

노션 클로닝 과제를 진행했습니다.
![ezgif com-video-to-gif](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/assets/43228743/36e8110b-1abc-448c-8ad3-c27ee206b880)

배포 링크: https://fedc-4-5-project-notion-vanilla-js-sigma.vercel.app/

- **webpack**을 사용하여 main.js를 진입점(entry point)으로 하여 모든 컴포넌트들을 불러오도록 했습니다. 
- eslint(airbnb-base), prettier를 구현하여 코드 일관성을 유지했습니다.
- **webpack**의 alias 기능을 사용해 import 문 길이를 최소화 하였습니다.
```
yarn install
npm install
```
라이브러리들을 설치하고 실행하면 프로젝트가 구동됩니다.
```
yarn start
npm start
```

### 컴포넌트 구조 
#### App.js
헤더와 콘텐츠 두 부분으로 컴포넌트를 나누었습니다.
![01](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/assets/43228743/a191e8fe-b34d-4fc0-96ff-cadf3b422fb5)


#### HeaderWrapper.js
![02](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/assets/43228743/c738c3c1-5dd7-4420-b4db-5ee6cf1e2885)

#### ContentWrapper.js
SideBarHeader 컴포넌트에 SideBar 컴포넌트를 열고 닫는 버튼을 만들었습니다.
![03](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/assets/43228743/d80974b5-1836-41ef-adf9-9af66c268e31)

#### SideBar.js
DocumentList, Document 컴포넌트가 서로 재귀적으로 호출하도록 구현하였습니다.
![04](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/assets/43228743/1c614ff4-3f2b-4f86-8d16-5d7fd2e91eb5)


## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->

### 기본 요구사항 구현 내용
- [x] 글 단위를 Document로 하고 Document는 Document를 여러개 포함할 수 있습니다.

- [x] 화면 좌측에 Root Documents를 불러오는 API를 통해 루트 Documents를 렌더링 합니다.
  - [x] Root Document를 클릭하면 오른쪽 편집기 영역에 해당 Document의 Content를 렌더링 합니다.
  - [x] 해당 Root Document에 하위 Document가 있는 경우, 해당 Document 아래에 트리 형태로 렌더링 합니다.
  - [x] Document Tree에서 각 Document 우측에는 + 버튼이 있습니다. 해당 버튼을 클릭하면, 클릭한 Document의 하위 Document로 새 Document를 생성하고 편집화면으로 넘깁니다.
  - [x] documentId에 해당하는 Document를 삭제합니다.
    - [x] 만약 하위 documents가 있는 document를 삭제한 경우, 하위 documents 등은 상위 document가 없어지므로 root document로 인식됩니다.  


- [x] 편집기에는 기본적으로 저장 버튼이 없습니다. Document Save API를 이용해 지속적으로 서버에 저장되도록 합니다.
  - [x] DocumentPage의 onEditing 함수에 디바운스를 적용하여 마지막 요청만 실행하도록 구현하였습니다.
- [x] History API를 이용해 SPA 형태로 만듭니다. 
  - [x] Root URL 접속 시엔 별다른 편집기 선택이 안 된 상태입니다.
  - [x] /document/{documentId}로 접속 시, 해당 Document의 content를 불러와서 편집기에 로딩합니다.
 
### 추가 구현 내용  
- [x] 사이드바 토글 버튼을 구현하여 여닫는 기능을 추가하였습니다.
- [x] 글이 저장된 것을 사용자에게 UI로 알려주기 위해 프로그래스 바를 추가하였습니다.
- [x] 현재 편집중인 Document가 저장되면, SideBar에도 적용되도록 구현하였습니다.
- [x] 현재 Document의 열림 / 닫힘 상태를 저장하기 위해 LocalStorage를 사용하였습니다.
- [x] 현재 documentID는 전역적으로 사용될 수 있어서 따로 유틸기능으로 구현했습니다. 

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- 컴포넌트를 너무 많이 나눈건 아닌가 하는 생각도 드는데 다들 어떻게 생각하시는 지 궁금합니다...
- 최대한 컴포넌트를 직접 렌더링 하지 않고, 상태만 변경하도록 코드를 작성했습니다. 혹시 그렇지 않은 부분이 있다면 가감없이 말씀해주시면 감사하겠습니다. 
- DocumentList와 Document가 서로를 호출하는 방식으로 구현되어있습니다. 어떤 부작용이 있는지는 잘 모르는 상태인데 문제점이 있다면 말씀 부탁드립니다.
- 성능 개선을 어떻게 해야할 지 잘 모르겠습니다... 비효율적이라고 생각되는 부분이 있다면 말씀 부탁드립니다! ㅠㅠ
- 이런 기능이 있으면 좋을것 같다고 생각하시는 것들도 말씀해주시면 감사하겠습니다.

wukdddang committed:
Init: 프로젝트 세팅
- Webpack Configuration
- Eslint(Airbnb) + Prettier

wukdddang committed:
Style: 폰트 적용

wukdddang committed:
Feat: 프로젝트 세팅 변경
- webpack alias 적용
- eslint에 webpack 플러그인 설정 적용

wukdddang committed:
Feat: Header 컴포넌트 추가

wukdddang committed:
Feat: Entry Point 추가

wukdddang committed:
Delete: 루트 디렉토리 style.css 삭제

wukdddang committed:
Feat: SideBar 컴포넌트 추가 (초안)

wukdddang committed:
Feat: webpack alias 추가

wukdddang committed:
Refactor: eslint 설정 변경

wukdddang committed:
Feat: request 기능 추가

wukdddang committed:
Feat: 문서 상태 저장을 위한 localStorage 기능 추가

wukdddang committed:
Feat: localStorage 상수 key 추가

wukdddang committed:
Feat: 라우팅 기능 추가

wukdddang committed:
Style: DocumentList 스타일 추가

wukdddang committed:
Feat: Document, DocumentList 컴포넌트 추가 - (1)

- Document, DocumentList 컴포넌트를 재귀적으로 호출하도록 구현
- documents가 배열이고, 길이가 1 이상이면 DocumentList 컴포넌트 호출
- onToggle, toggleDocument 함수로 현재 문서 상태 localStorage에 저장
- onAdd 함수로 새로운 문서 자동 추가

wukdddang committed:
Refactor: webpack alias 추가 및 변경

wukdddang committed:
Refactor: App.js 클래스명 변경

wukdddang committed:
Chore: package.json url-loader 제거

wukdddang committed:
Feat: yarn-error.log 추가

wukdddang committed:
Chore: 프로젝트 시작 파일들 수정

wukdddang committed:
Feat: App.js 추가
- routing 기능 구현
- documentId값을 가지면 contentWrapper의 상태를 업데이트

wukdddang committed:
Feat: DirectoryPage, SideBar, SideBarHeader 컴포넌트
- 사이드바 여닫는 기능은 DirectoryPage 컴포넌트에서 처리하도록 구현

wukdddang committed:
Style: SideBar 컴포넌트 스타일 추가

wukdddang committed:
Style: SideBarHeader 컴포넌트 스타일 추가

wukdddang committed:
Style: DirectoryPage 컴포넌트 스타일 추가

wukdddang committed:
Style: 앱 기본 스타일 추가
- 웹 폰트: Pretendard

wukdddang committed:
Feat: API request 함수 구현

wukdddang committed:
Feat: Document 상태 유지를 위한 localStorage 기능 추가

wukdddang committed:
Feat: 사이드바 닫기/열기용 유틸 함수 추가

wukdddang committed:
Feat: URL 라우팅 이벤트 처리 기능 추가

wukdddang committed:
Feat: 전역 documentId 상태 저장을 위한 유틸 기능 추가

wukdddang committed:
Delete: 폴더 디렉토리 구조 변경

wukdddang committed:
Feat: API URL 상수로 처리하는 기능 추가

wukdddang committed:
Feat: 문서 저장 표현을 위한 프로그레스 바 구현

wukdddang committed:
Style: ProgressBar 컴포넌트 스타일 추가

wukdddang committed:
Feat: Header 컴포넌트 추가

wukdddang committed:
Feat: HeaderWrapper 컴포넌트 추가

wukdddang committed:
Feat: TextEditor 컴포넌트 추가

wukdddang committed:
Feat: ContentWrapper 컴포넌트 추가

wukdddang committed:
Feat: DocumentList 컴포넌트 구현
- Document 컴포넌트에서 documents 요소가 있다면 재귀적으로 호출하도록 구현

wukdddang committed:
Feat: Document 컴포넌트 추가
- Document state 내부에 documents 배열이 존재하는 경우 DocumentList 컴포넌트 렌더링
- 20자 이상 시 ...으로 문장 줄임 표현

wukdddang committed:
Refactor: TextEditor 오류 수정
- 저장 시 편집기의 상태도 업데이트 하도록 수정

wukdddang committed:
Feat: DocumentPage 컴포넌트 추가
- 저장완료를 알리는 ProgressBar 상태 업데이트

wukdddang committed:
Refactor: TextEditor 컴포넌트 수정
- console.log 제거

wukdddang committed:
Refactor: Document 폴더 DocumentList 폴더 내부로 이동

arclien commented on .eslintrc.js:
```diff
@@ -0,0 +1,26 @@
+      },
+    },
+  },
+  rules: {
+    'no-new': 0, // new 연산자를 사용하고 할당하지 않아도 경고하지 않는다.
+    'no-unresolved': 0, // import 구문에서 경로를 검사하지 않는다.
+    'prettier/prettier': 'error', // prettier의 문법 검사 결과를 오류로 처리한다.
+    'import/no-unresolved': 2, // import 구문에서 경로를 검사한다.
+    'import/no-cycle': 0, // 재귀 참조를 허용한다.
+    'no-restricted-globals': 0, // window.alert 같은 내장 전역 변수를 사용할 수 있다.
+  },
```
굿굿!!
같이 협업하는 동료마다 해당 rule은 같이 정해야하지만, 각각을 잘 알고 있으면
창욱님이 직접 건의해서 추가하기도 수월할 것 같습니다!! 👍

wukdddang replied:
감사합니다! ㅎㅎ 🙇🏻‍♂️ eslint 설정을 하나씩 알아가는 재미도 있는 것 같습니다... ㅎㅎ

arclien commented on .prettierrc.js:
```diff
@@ -0,0 +1,10 @@
+  arrowParens: 'always',
+  bracketSpacing: true,
+  endOfLine: 'auto',
+}
```
EOF 지켜주세요!

재밌는(흥미로운) 점은, 호원님, 혜진님의 prettierrc 세팅이 다다르다는 것이네요!
정답이 있는게 아니라서 재밌다고 표현했습니다!

[호원님 세팅](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/45/files#diff-979a4a52db5287c891eac28a141bfa97bf8b0732ed5afae848d90956ef80e78d)
[혜진님 세팅](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/36/files#diff-7312296367dd7c725b9beb500d7dab7bbeeb2aa03a1010a3fb2935f826d03fa6)

Lim-JiSeon replied:
저번에 prettier 조언해주셔서 감사합니다. 저도 덕분에 수정해서 좀 더 자연스러운 코드를 작성할 수 있었어요. 
멘토님 말씀대로 다들 다르게 세팅한 점이 재미있는 것 같아요 ㅎㅎ

wukdddang replied:
이번 기회에 EOF에 대해 더 자세하게 살펴볼 수 있었던 것 같습니다. ㅎㅎ
vscode 상의 editorconfig 기능과 prettier를 사용해서 EOF를 해결하도록 했습니다! 감사합니다 ㅎㅎ ☺️☺️

arclien commented on index.html:
```diff
@@ -0,0 +1,11 @@
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Document</title>
```
창욱님께서는 프로덕션 배포도 하셨으니,
meta tag에 대해서도 공부해보시죠!

documents를 라우팅시마다 메타태그를 업데이트 해주면 더 좋겠죠! => SEO에도 좋고요!

https://www.w3schools.com/tags/tag_meta.asp

wukdddang replied:
넵 멘토님, 노션 리팩토링 스터디 진행하면서 meta 태그도 업데이트 해보겠습니다. 감사합니다!! ☺️🙇🏻‍♂️🙇🏻‍♂️

arclien commented on src/style.css:
```diff
@@ -0,0 +1,44 @@
+  font-style: normal;
+}
+
+body {
+  font-family: 'Pretendard', sans-serif;
+  margin: 0;
+  padding: 0;
+}
+
+* {
+  box-sizing: border-box;
+  list-style: none;
+}
+
+li,
+ul,
+span {
+  margin: 0;
+  padding: 0;
+}
+
+h1,
+h2,
+h3,
+h4,
+h5,
+h6 {
+  margin: 0;
+}
```
브라우저 기본 스타일을 override 하는 경우 reset.css로 관리하는 편입니다!
reset.css, normalize.css 등으로 검색해보시면 좋을 것 같아요!

arclien commented on src/utils/global/index.js:
```diff
@@ -0,0 +1,7 @@
+let currentDocumentId = null;
+
+export const getCurrentDocumentId = () => currentDocumentId;
+
+export const setCurrentDocumentId = (documentId) => {
+  currentDocumentId = documentId;
+};
```
아직 코드를 다 읽지 않은 상태에서 질문입니다.
localStorage가 아니고, currentDocumentId 변수에 상태를 저장한 이유가 무엇인가요?

wukdddang replied:
현재 선택된 문서가 무엇인지 UI로 표현하기 위해, 전역에서 문서의 id를 기억하고자 했습니다.
그런데 이미 url로 현재 문서의 id를 알고 있으니 굳이 currentDocumentId에 저장할 필요가 없다는 것을 지금 깨닫게 되었습니다... 🙇🏻‍♂️

arclien commented on src/utils/sidebarUI/index.js:
```diff
@@ -0,0 +1,13 @@
+let isOpen = true;
+
+export function openSideBar() {
+  isOpen = true;
+}
+
+export function closeSideBar() {
+  isOpen = false;
+}
+
+export function getSideBarState() {
+  return isOpen;
+}
```
src/utils/global/index.js 와 유사한 것 같습니다.

sidebarUI라서 분리하신 의도는 알 것 같은데, 이것 또한 global하게 관리되어 어디서든 접근이 가능한 것 같아요.
그래서 global 폴더 하위에 index.js 에서 documentId와 sidebarOpen 상태를 같이 관리하거나
두개의 파일 혹은 하위 폴더로 관리하는게 좋을 것 같아요.


혹은 globalState라는 객체를 갖고 거기에서 currentDocumentId, isSidebarOpen 이렇게 프로퍼티로 관리해도 좋을 것 같습니다

wukdddang replied:
넵 멘토님, 리뷰 감사합니다! 🙇🏻‍♂️🙇🏻‍♂️
다른 분들의 리뷰를 종합했을 때... 현재 사이드바와, 프로그래스 바가 state의 흐름을 방해하는 것 같아, 전체적인 앱 구성을 변경해야할 것 같습니다... 그래서 이번주 리팩토링 시간에 애플리케이션 구성을 변경하고, Sidebar의 state가 열고 닫는 상태를 가질 수 있도록 변경하도록 하겠습니다!!

arclien commented on src/utils/storage/index.js:
```diff
@@ -0,0 +1,43 @@
+const storage = window.localStorage;
+
+const getOpenDocumentIds = (openDocuments, documents) => {
```
여기에서 도큐먼트의 getOpen, setOpen 상태를 localStorage로 저장하고 계시네요!

처음에는 열려있다는 의미라서, documents/${documentId} 접근해서 해당 도큐먼트를 작업중(보는중)이라는 의미의 open으로 생각했습니다.

그게 아니라, sidebar에서 toggled(activated) 상태임을 나타내는 것 같은데, 조금 더 시멘틱한 함수명을 주면 좋을 것 같습니다! :)

wukdddang replied:
넵 멘토님 확실히 코드명이 모호한 것 같습니다...!
`getOpenDocumentIds`는 `getToggledDocumentsIds`로, `getDocuments`는 `getToggledDocuments`로 변경했습니다!
리뷰 감사합니다!! 🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️ [관련 커밋](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7/commits/472f0b2d37ccf34aa98b781d30d0336893568374)

arclien commented on src/components/HeaderWrapper/Header/style.css:
```diff
@@ -0,0 +1,5 @@
+.Header {
+  font-size: 4vw;
+  font-family: 'Pretendard', sans-serif;
```
src/styles.css에서 
이미 body에 font를 정의해서 추가 정의 하지 않아도 될 것 같아요!

wukdddang replied:
넵 맞습니다!! ㅠㅠ 수정하도록 하겠습니다! 🙇🏻‍♂️

arclien commented on src/components/HeaderWrapper/index.js:
```diff
@@ -0,0 +1,15 @@
+
+import './style.css';
+
+export default function HeaderWrapper({ $target, initialState }) {
```
선호도 차이일 수 있습니다!

컴포넌트 이름에서 Wrapper를 잘 사용하지 않는 편입니다.(아직 사용 예제를 본 적도 없습니다 )
Header 컴포넌트로 우선 짓고,
하위 컴포넌트를 Header/ProgressBar, Header/Title 이렇게 두개로 나눠서 갖는건 어떨까요?
```
components
   - Header( 기존 HeaderWrapper )
       - Title ( 기존 Header)
       - ProgressBar

```

kutta97 replied:
저도 이 부분에 대해 조금 이상함을 느껴서 리뷰를 달러 왔더니, 이미 멘토님의 리뷰가 있으셨네요ㅎㅎ

저는 이렇게 기능이 있는 컴포넌트 이름에는 보통은 Wrapper 를 잘 사용하지 않는다고 알고 있습니다. 다만, 정말 Wrapper 에 있는 기능이 그 자체로서 자주 사용되어 재사용이 필요한 경우에 간혹 Wrapper 컴포넌트를 만드는 경우도 있습니다. 하지만 이 경우에는 지석멘토님의 말씀처럼 Header 컴포넌트로 이름을 변경하고, Header 컴포넌트를 Title 로 바꾸는 것이 더 자연스러울 것 같습니다.

일반적인 경우에 wrapper, container 등의 이름은 보통 css classname 과 같이, '논리적'인 구분이 아닌, '디자인'에 의한 구분을 표현하기 위해 붙이는 이름으로 사용하는 것이 좋다고 생각합니다.

wukdddang replied:
이번 리팩토링 스터디 때 대대적인 수정을 기하도록 하겠습니다.... 
리뷰 너무 감사합니다!! ☺️☺️ 🙇🏻‍♂️

arclien commented on src/components/ContentWrapper/index.js:
```diff
@@ -0,0 +1,28 @@
+import DocumentPage from '@components/ContentWrapper/DocumentPage';
+import DirectoryPage from '@components/ContentWrapper/DirectoryPage';
```
ContentWrapper 하위에서
DirectoryPage(사이드바), DocumentPage( 우측 text editor)를 렌더링하고 있습니다.
저는 Page라는 postFix는 정말 URL에 매칭되는 제일 상위 Page를 뜻할때 사용하곤 합니다.

components 폴더 하위에 Page가 있는게 살짝 어색해보이는 것 같아요.
컴포넌트는 의미있는  작은 단위를 의미하고, page는 components의 조합을 통한 페이지를 나타내는 것 같다고 생각합니다.

wukdddang replied:
넵 멘토님, 이번 리팩토링 스터디 때 전체적인 파일 명을 변경하고 프로젝트 구성도 변경해보려고 합니다.. (Sidebar 컴포넌트 에서 API 호출이 아닌 APP 컴포넌트 에서 호출)
리뷰 너무 감사합니다!! ☺️☺️ 🙇🏻‍♂️

arclien commented on src/components/ContentWrapper/DocumentPage/index.js:
```diff
@@ -0,0 +1,61 @@
+    this.state = nextState;
+    this.render();
+  };
+  const $progressBar = document.querySelector('.ProgressBar');
```
$progressBar 가 존재하는지 혹은 null를 반환하는지 체크해보면 좋을 것 같습니다.
다음코드에서 $progressBar.style.opacity  `Cannot read properties of null`를 반환할지도 모릅니다.(물론 지금은 잘 되고있습니다!)

ProgressBar 식별자 또한 constant로 관리하면 좋을 것 같습니다!

kutta97 replied:
`ProgressBar`는 `HeaderWrapper` 에 존재하는 컴포넌트인 것 같습니다..!

`state` 관리를 `ContentWrapper` 에서 하게 된다면, state 가 단방향으로만 흐르지 않게 됩니다. 지금은 ProgressBar 컴포넌트 하나만 찾아서 상태를 바꾸고 있기 때문에 괜찮지만, 프로젝트가 복잡해질 수록 자식 컴포넌트가 다른 컴포넌트를 찾아 상태를 바꾸는 방식은 코드의 흐름을 이해하고, 유지보수하기 힘들어질 가능성이 높습니다..

ex) 사이드 바의 document 를 눌렀을 때 로딩바를 보여줘야 한다면?, document 페이지의 하위 링크를 눌렀을 때 로딩바를 보여줘야 한다면?? 이런 경우마다 일일히 `ProgressBar`를 querySelector 로 찾아서 state 를 바꿔준다면, 다른 사람이 코드를 읽을때는 어느 시점에 프로그래스바의 상태가 변경되는지 이해하기 어려워질 것입니다.

isLoading 변수를 Page, 또는 전역에서 가지고 있고, API 등의 비동기 호출이 있을 때마다 isLoading 상태를 변경하여, ProgressBar 가 움직이도록 하는 것이 더 좋을 것 같다고 생각합니다.

wukdddang replied:
리뷰 감사드립니다! 🙇🏻‍♂️🙇🏻‍♂️ 현재 혜진님이 지적해주신대로, 프로그래스 바를 동작할 때 state가 단방향으로 흐르고 있지 않고 ContentWrapper에서 독자적으로 컴포넌트의 상태를 바꿔주고 있습니다. 전역에서 isLoading 변수를 가지면서 저장되는 API 호출이 생길 때 마다 state를 변경하기 위해서는 전체적인 애플리케이션의 구성 변경이 필요할 것으로 생각되어 시간이 좀 걸릴 것 같습니다.. 
(시간이 걸릴것 같다고 생각하는 이유 1. Sidebar에서 API 호출이 일어나고 있는데, Sidebar 컴포넌트에 render 함수가 없음... 구현이 필요함)
(시간이 걸릴것 같다고 생각하는 이유 2. Sidebar는 this.render함수를 가지고 있지 않고 Sidebar의 this.setState()에서 documentList.setState()가 일어나고 있음. this.render()로 이동이 필요함)

이번주 리팩토링 기간에 
1. DOM 객체의 null 여부 체크와, 
2. 단방향 state를 가질 수 있도록 기능 구현하도록 하겠습니다!

arclien commented on src/components/ContentWrapper/DocumentPage/index.js:
```diff
@@ -0,0 +1,61 @@
+          $progressBar.style.width = '100%';
+
+          textEditor.setState({ status: 'saved', ...createdDocument });
+          updateState({ status: 'saved', ...createdDocument });
```
createdDocument 가 실제로 존재하는지 체크하는 로직을 추가하면 좋을 것 같아요!
백앤드 실수로 데이터가 안내려올 수 있으니까요!

wukdddang replied:
넵 멘토님. 만약 res가 없다면, error를 출력할 수 있도록 방어 코드를 추가했습니다! [링크](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7/commits/e193b82778e58dd7a0f026dc9060aeeca77ff1cd)
리뷰 감사합니다! ㅎㅎ ☺️

arclien commented on src/components/ContentWrapper/DocumentPage/index.js:
```diff
@@ -0,0 +1,61 @@
+          $progressBar.style.width = '100%';
+
+          textEditor.setState({ status: 'saved', ...createdDocument });
+          updateState({ status: 'saved', ...createdDocument });
```
status: 'saved'를 사용하셔서, TextEditor에서 status를 관리하는 것 같았으나 그게 아니네요!
지금의 프로퍼티명(status)는 약간의 혼란을 주는 것 같아요!
어떠한 status에 따라 로직 처리를 할 것 같은데, 실제로는 TextEditor에서 null 체크만 하고있어서요

boolean 형태로 바꿔보는 것은 어떨까요?

wukdddang replied:
넵 멘토님 말씀대로, status: string으로 표현되는 state를 isDocumentSaved: boolean 형태로 변경했습니다. [링크](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7/commits/1ea26a5dcde1cff7fb85bc68b83e1bfa685a3b30)
리뷰 감사합니다! ㅎㅎ ☺️☺️

arclien commented on src/components/ContentWrapper/DirectoryPage/SideBar/index.js:
```diff
@@ -0,0 +1,91 @@
+    documentList.onRemove(documentId);
+  };
+
+  $sideBar.addEventListener('click', ({ target }) => {
+    const $documentHeader = target.closest('.DocumentHeader');
+    if (!$documentHeader) return;
+
+    const { documentId } = $documentHeader.dataset;
+
+    if (target.classList.contains('DocumentTitle')) {
+      handleDocumentTitleClick(target);
+    }
+
+    const $documentAddButton = target.closest('.DocumentAddButton');
+    const $documentToggleButton = target.closest('.DocumentToggleButton');
+    const $documentDeleteButton = target.closest('.DocumentDeleteButton');
+
+    if ($documentAddButton) {
+      handleDocumentAddButtonClick(documentId);
+    } else if ($documentToggleButton) {
+      handleDocumentToggleButtonClick(documentId);
+    } else if ($documentDeleteButton) {
+      handleDocumentDeleteButtonClick(documentId);
+    } else {
+      push(`/documents/${documentId}`);
+    }
```
창욱님!! 정말 이벤트 바인딩을 잘 하신 것 같습니다!
1. 이벤트 위임도 잘하셨으며
2. 각각의 로직에 대해서 추상화를 해서 실제 코드는 다른 곳에서 처리를 하고, 여기에서는 각각의 조건일 때 어떠한 의미를 갖는 함수를 호출하는지만 파악할 수 있어서 좋습니다!
👍 👍 👍 👍 👍

kutta97 replied:
오 이렇게 이벤트 바인딩 처리를 할 수 잇군요!
저도 창욱님 코드에서 배워갑니다!!!

Lim-JiSeon replied:
멘토님과 혜진님 말씀대로 정말 코드를 잘 짜신 것 같아요!
가독성이 좋아서 바로 이해할 수 있는 쉬운 코드였습니다 💯 
많이 고민하신 것 같아 저도 자극받네요 👍

hayamaster replied:
와....이벤트 바인딩을 찢어놓으셨다....💥 
배우겠습니다 창욱님

wukdddang replied:
부족한 코드 좋게 봐주셔서 너무 감사합니다. ☺️

arclien commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/index.js:
```diff
@@ -0,0 +1,66 @@
+  };
+
+  this.onAdd = async (documentId) => {
+    const res = await request('/documents', {
+      method: 'POST',
+      body: JSON.stringify({ title: '새로운 문서', parent: documentId ? Number(documentId) : null }),
+    });
+    push(`/documents/${res.id}`);
```
400에러가 아니라, 200으로 떨어졌지만 혹시라도 백앤드 이슈로
res에 id값이 없을 경우를 대비해 방어코드를 추가하면 좋겠습니다!

wukdddang replied:
넵 멘토님. 만약 res가 없다면, error를 출력할 수 있도록 방어 코드를 추가했습니다! [링크](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7/commits/affb522c4a2bbdb67c89b232219f1ec764fdc0e5)
리뷰 감사합니다! ㅎㅎ ☺️

arclien commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/index.js:
```diff
@@ -0,0 +1,49 @@
+      $documentContent.innerText = '하위 문서가 없습니다. 🤔';
+    }
+
+    const displayedTitle = title.length > 20 ? `${title.substring(0, 20)}...` : title;
```
마이너한 사항인데, 20 과 같은 상수 값은 따로 뺴두면 좋을 것 같아요!
지금은 여기 한줄에서만 쓰이지만, 로직이 복잡해져서 여러곳에서 쓰이면 휴먼에러가 나올 수 있습니다!

Lim-JiSeon replied:
문서 제목이 길어질 경우까지 고려하신 점이 좋은 것 같아요! 
사용자 측면에서 다양하게 세심하게 고려하시고 구현하신 부분이 잘 드러난 것 같습니다 ㅎㅎ
하위 문서가 없을 경우 나타나는 텍스트도 재미있었어요 💯

wukdddang replied:
일단 기능을 구현하고 보자는 식으로 만들다 보니 매직 넘버라는 것을 고려하지 못했습니다...
이 부분은 이번주 리팩토링할 때 애플리케이션 구조를 변경하면서 바꾸도록 하겠습니다! 리뷰 감사합니다 🙇🏻‍♂️🙇🏻‍♂️

arclien reviewed (commented):
창욱님!! 정말 나날이 발전하는게 보입니다!
프로젝트 세팅, 배포, PR 문서 작성, 코드 모두 다 좋습니다!! 👍 👍 👍 

아래 몇 가지 의견 드리니 참고해주세요!

1. 최초 사용자의 경우 최초 문서를 만들 방법이 없어서, newbutton이 필요할 것 같습니다.
(`'x-username': 'wukddang'`  이 값을 일부러 다른 임의의값으로 변경해보시면 알 겁니다! )
3. https://fedc-4-5-project-notion-vanilla-js-sigma.vercel.app/documents/99999999999  이렇게 말도 안되는 documentId값이 들어올 경우 방어를 하면 좋을 것 같습니다!
4. style.css 파일을 각각의 컴포넌트에서 만들어주셨는데요!! 나중에 style.css 파일 이름 검색하면 너무 많은 검색이 되어서, 가능하면 컴포넌트 이름과 매칭시켜서 css 파일 이름을 갖게 되면 어떨까 싶습니다! ( 이건 호불호의 차이일 수 있습니다!)
5. 컴포넌트는 잘 분리하신 것 같아요! 단지 폴더 구조(depth)가 조금 많은 것 같고 page/component 혼용이 있는 것 같아요. 컴포넌트의 경우 잘게 쪼개더라도 하위 depth로 넣지 말고, 의미적으로 같은 레벨의 하위 컴포넌트라면 같은 level에 존재해도 좋다고 저는 생각합니다. 
아, 그런데 모든 컴포넌트를 index.js를 통해서 export 해서 sub folder 구조가 나올 수 밖에 없겠군요!!
제가 위에했던 말에서는, documentItem.js, documentList.js 이렇게 두개가 필요시 같은 레벨에 존재해도 될 것 같다는 의미긴 했습니다!

수고하셨습니다!

howons commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/index.js:
```diff
@@ -0,0 +1,49 @@
+    $documentContent.style.display = isOpen ? 'block' : 'none';
+
+    if (Array.isArray(documents) && documents.length > 0) {
+      new DocumentList({ $target: $documentContent, initialState: documents }).setState(documents);
```
생성 후 바로 setState를 해줄 수 있다는 건 생각치 못했네요.👍

wukdddang replied:
저도 이번에 알게 되었습니다. ㅎㅎ 한줄로 만들 수 있을 것 같은데... 고민하다보니 이런 방법이 있더라구요 ☺️

kutta97 commented on src/App.js:
```diff
@@ -0,0 +1,30 @@
+  this.route = async () => {
+    const { pathname } = window.location;
+
+    if (pathname === '/' || pathname === '/documents') {
```
우선, documents/ 뒤에 페이지 id 가 들어오지 않은 경우의 예외처리를 해주신 부분은 좋다고 생각합니다. 개발자에게 예외처리를 꼼꼼하게 하는 능력이 중요하다고 생각합니다.

다만, url 이 `'/documents'` 로 들어온 경우는 과제 요구사항 명세에 없었으므로, 조건문에서 빼주시는 것이 좋을 것 같다고 생각합니다!

사전에 정의되지 않은 url 로 접속한 경우(404 Error)에는 root 페이지로 이동하도록 처리하는 것이 더 자연스러워 보입니다.

wukdddang replied:
"/documents url이 있어야 하지 않을까?" 하고 막연하게 생각하고 있다보니 그냥 추가해버렸습니다... 넵 수정하는 게 좋아보입니다 ☺️ 리뷰 감사합니다!! 🙇🏻‍♂️

kutta97 commented on src/App.js:
```diff
@@ -0,0 +1,30 @@
+    } else if (pathname.indexOf('/documents/') === 0) {
+      const [, , id, status] = pathname.split('/');
+      const res = await request(`/documents/${id}`);
+      setCurrentDocumentId(id);
```
이건 제 개인적인 궁금함인데요, `documentId` 를 전역 변수로 저장하신 이유가 궁금합니다!
url 에도 해당 정보가 있고, `contentWrapper` 컴포넌트에도 state 로 documentId 데이터를 내려주고 있는것으로 보여서요..!

wukdddang replied:
"window.location을 이용해서 id를 가져오는게 낫겠다." 고 생각했다가, "그러면 id가 필요할 때마다 코드가 반복되지 않을까...?" 하고 막연하게 사고의 흐름을 마무리하고 전역 변수에 저장하자는 결론을 내렸습니다. currentDocumentId의 경우 Sidebar에서 현재 선택된 문서를 UI적으로 알려주는 용도로 사용하려고 했습니다.

문제는 App 에서 API 호출이 이뤄지지 않고... Sidebar에서 API 호출이 이뤄지고 있습니다. Sidebar 컴포넌트를 렌더링 할 때, 문서 데이터들이 필요하니까 굳이 App에서 하지 않고 Sidebar에서 호출했는데, 이게 App에서 전달되는 state를 유지하지 못하게 하는 로직이었던 것 같습니다... 
그러다보니 부득이 전역 상태를 가지도록 했는데 굳이 전역을 가질 필요는 없어보이고, window.location 등을 이용해서 id를 확인하도록 하면 될 것 같습니다. 이 부분은 수정하도록 하겠습니다!!! 🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️

번외로, 혜진님은 현재의 state가 다뤄지는 방식을 어떻게 생각하시는 지 궁금합니다. App에서 호출을 하는 게 좋을까요? 아니면 Sidebar에서 렌더링을 하니까 Sidebar가 가지는 게 좋을까요? 

그런데 평소 혜진님이 생각하시는 바로 유추해보면.... App에서 state를 계속 전달하는 게 데이터 흐름 상 이해하기 쉬우니 전자를 선호하실 것 같습니다... 어렵네요... 고민하면서 저도 많이 배운 것 같습니다. 감사합니다 ㅎㅎ ☺️☺️

howons commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/index.js:
```diff
@@ -0,0 +1,66 @@
+    push(`/documents/${res.id}`);
+  };
+
+  this.onRemove = async (documentId) => {
+    const res = await request(`/documents/${documentId}`, {
+      method: 'DELETE',
+    });
+    if (res.parent && res.parent.id) {
+      push(`/documents/${res.parent.id}`);
+    } else {
+      push('/documents');
+    }
+  };
```
api 호출에서 가지는 관심사는 컴포넌트의 관심사랑 별개라고 생각해서 따로 분리하면 좋을 것 같아요. 🙂
보통 api폴더에 분리하는 것 같아요. [출처](https://ghost4551.tistory.com/163)

wukdddang replied:
이번주 리팩토링 하면서.. 한번 시도해보겠습니다! 생각보다 고쳐야 할 게 너무 많아서 정신이 없네요 🥲

howons commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/index.js:
```diff
@@ -0,0 +1,66 @@
+    const nextState = this.state.map((document) => toggleDocument(document, documentId));
+    setDocuments(LOCAL_STORAGE_KEY, nextState);
+    this.setState(nextState);
+    push(`/documents/${documentId}`);
```
제 성향일 수도 있지만 토글 버튼은 하위 문서 뭐 있나 보는 용도로 쓰는 것 같습니다.
토글 버튼 클릭할 때 url은 push 하지 않아도 괜찮을 것 같아요!

wukdddang replied:
너무 정확한 지적이십니다. UX적으로 너무 아쉬운 코드였네요... 바로 수정했습니다. ㅎㅎ 감사합니다!! ☺️☺️ 🙇🏻‍♂️🙇🏻‍♂️

howons commented on src/utils/router/index.js:
```diff
@@ -0,0 +1,19 @@
+    history.pushState(null, null, nextUrl);
+    onRoute();
+  });
+};
```
`popstate`이벤트 처리를 깜빡하신 것 같습니다!
뒤로가기나 앞으로가기 시에 edit 페이지가 동기화되지 않고 있는 것 같아요.

wukdddang replied:
넵 호원님! 짚어주셔서 감사합니다 ㅎㅎ
고치면서, 또 애플리케이션의 좋지 않은 점을 발견해서 슬픈 하루입니다... 🥲 
리뷰 감사합니다!!! 🙇🏻‍♂️🙇🏻‍♂️ (링크)[https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7/commits/c058834ed376e546c11064c7d9afa983a06f8f1c]

kutta97 commented on src/components/ContentWrapper/DirectoryPage/SideBar/index.js:
```diff
@@ -0,0 +1,91 @@
+    documentList.onAdd(documentId);
+  };
+
+  const handleDocumentToggleButtonClick = (documentId) => {
+    documentList.onToggle(documentId);
+  };
```
documentList.onToggle(documentId); 이 함수에 ui 를 toggle 하는 것 뿐만 아니라,  해당 id 의 페이지로 이동하는 side effect 가 있는 것 같습니다!

저는 현재 페이지 상태는 유지하면서, Sidebar 의 document item 을 ui 만 toggle 하고 싶은데, document item 을 toggle 할때마다 해당 item 의 페이지로 이동하는 버그가 있는 것 같습니다.

이 부분은 버그인 것으로 보입니다. 고쳐주시면 좋을 것 같습니다.

Lim-JiSeon replied:
this.onToggle = (documentId) => {
    const nextState = this.state.map((document) => toggleDocument(document, documentId));
    setDocuments(LOCAL_STORAGE_KEY, nextState);
    this.setState(nextState);
    push(`/documents/${documentId}`);
    
코드를 보면 해당 item 페이지 링크를 push 하는데 이 때문인 것 같습니다. 
만약 창욱님이 의도하셨던 바가 아니시라면 혜진님 말씀대로 수정하면 좋을 것 같습니다.

wukdddang replied:
버그가 맞습니다... 빠른 시일내에 수정하도록 하겠습니다 🙇🏻‍♂️

wukdddang replied:
수정 완료했습니다 ㅎㅎ 감사합니다! ☺️☺️

kutta97 commented on src/components/ContentWrapper/DirectoryPage/index.js:
```diff
@@ -0,0 +1,40 @@
+    this.render();
+  };
+
+  this.render = () => {
+    if (getSideBarState()) {
+      sideBar.show();
+    } else {
+      sideBar.hide();
+    }
```
sidebar 를 보여주고, 숨겨주는 로직을 직관적인 함수명으로 잘 표현해주신 것 같습니다.
하지만, component 는 state 를 기반으로 rendering 한다는 규칙이 있으니, 이 규칙의 일관성을 따라서, sidebar 의 show & hide 로직또한 state 로 관리해주는 것이 통일성 있을 것 같습니다.

show()라는 함수명은 직관적이어서 이 부분은 유지하면서 이런 느낌으로 변경해주시면 더 좋을 거 같습니다.

```javascript
const showSidebar = () => {
  sidebar.setState({ isVisible: true })
}
```

hayamaster replied:
혜진님 말씀 너무 맞는 말씀이십니다!!!
일관성있게 작성하는 것은 매우 중요한 것 같습니다!

다만 저는 창욱님께서 이렇게 구현하신 것도 직관적으로 잘 이해되는 것 같습니다! 
(지극히 개인적인 의견입니다...!)

wukdddang replied:
혜진님 말씀이 맞습니다! 반박할 말들이 떠오르지 않습니다 ㅎㅎ.
다만 현재 Sidebar 컴포넌트에서 API 요청을 하고 있었네요... 그리고 Sidebar 컴포넌트에서 this.render를 정의하지 않았습니다.. 그러다보니 흐름이 중간에 좀 이상해진 느낌이 있어서 애플리케이션 구조를 뜯어고쳐야 할 것 같습니다 😭

이번주 리팩토링 시간에 구조를 바꾸고, state를 단방향으로 흐를 수 있게끔 수정하고 state를 기반으로 rendering할 수 있도록 수정하도록 하겠습니다! 리뷰 감사합니다 🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️

kutta97 commented on src/components/ContentWrapper/DocumentPage/TextEditor/index.js:
```diff
@@ -0,0 +1,42 @@
+      ...this.state,
+      ...nextState,
+    };
+    // console.log(this.state);
```
사용하지 않는 코드는 특별한 이유가 없다면, 제거하는 습관을 들이시면 좋습니다!

Lim-JiSeon replied:
console.log로 한 번씩 확인해보는 거 개발자들 국룰인 거 같아요 ㅎㅎ
저도 많이 실수하는 부분인데 제 현재 코드에도 test용으로 classname을 1 2 로 설정한게 남아서 부끄럽네요.. 
저도 이 부분 주의해야겠어요 :)

wukdddang replied:
치부를 들켰네요... ㅎㅎ 수정 완료했습니다! 감사합니다! ☺️☺️

kutta97 commented on src/components/ContentWrapper/DocumentPage/TextEditor/index.js:
```diff
@@ -0,0 +1,42 @@
+  };
+
+  this.render = () => {
+    const { title, content, res, status } = this.state;
```
이 status 는 어떤 용도로 사용되는 변수인지 궁금합니다.
Http request status 를 의미하는 것인가요?  변수 이름이 조금 더 명시적으로 표현되었다면 좋을 것 같습니다.

wukdddang replied:
문서가 저장되었는지를 나타내는 변수였습니다! 확실히 status로 표현하니 의도가 명확하지 않아서 아쉬웠습니다...
수정했습니다! 리뷰 감사합니다 ☺️☺️

kutta97 commented on src/components/ContentWrapper/DocumentPage/TextEditor/index.js:
```diff
@@ -0,0 +1,42 @@
+
+  const onKeyUp = () => {
+    const { id } = this.state.res;
+    onEditing(id, { title: $titleEditor.value ? $titleEditor.value : '제목 없음', content: $textEditor.value });
+  };
```
사용자가 title 에 빈 문자열을 작성하기를 원하는 경우에도 '제목 없음'이라는 제목으로 document 가 저장되는 로직은 사용성 측면에서 조금 어색한 것 같습니다.

실제 Notion 앱처럼, 빈 문자열 제목을 허용하되, 만약 빈 문자열 제목이면 UI 만 'untitled'라고 보여지도록 수정하는 것은 어떨까요?

Lim-JiSeon replied:
저도 제목 없음은 조금 어색할 것 같아서,,, 빈 문자열을 허용하지 않거나, 다른 문자열을 보여주는 방식으로 수정하면 더 좋은 서비스가 될 것 같습니다~
혜진님 말씀대로 Untitled도 방법이 될 수 있고 다른 서비스들을 참고해보시면 좋을 것 같아요 :)

wukdddang replied:
빈틈이 많은 애플리케이션이네요 ㅎㅎ... 일단 기능구현에 목적을 맞추다 보니 구조에서 허술한 부분이 많은 것 같습니다.
해당 기능도 이번주 리팩토링 기간 마저 수정하도록 하겠습니다 리뷰 감사합니다! 🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️

howons reviewed (commented):
우리 조의 연막 쌍두마차 창욱님! 
너무 잘 만드셨고 코드에도 여러가지 고민의 흔적이 드러나는 것 같습니다. 코드도 읽기가 편해서 막힘 없이 리뷰할 수 있었어요. 배포에 대한 팁도 제일 먼저 공유해주셔서 덕분에 vercel 배포를 저도 시도해보는 계기가 되었습니다! 😌


다만 멘토님이 말씀하신 것처럼 폴더의 depth가 약간 많다고 느껴집니다.
하필 js 파일 이름도 모두 index라서 깃헙의 `Files changed` 탭에서 왔다갔다하기 혼란스러웠네요 😵‍💫

kutta97 reviewed (commented):
창욱님, 노션 클로닝 과제 하시느라 수고 많으셨습니다!
과제하면서 밤 샐 때마다 거의 대부분 창욱님도 계셔서... 동지(?)처럼 느껴져 의지가 되고 좋았던 것 같습니다.ㅎㅎ
더 자세히 분석하고 꼼꼼하게 리뷰 해드리고 싶은데, 코드 리뷰 마감일이 얼마 남지 않아서 코드를 얕게 봐드린것 같아 죄송하네요ㅠㅠ

### Q. 컴포넌트를 너무 많이 나눈건 아닌가 하는 생각도 드는데 다들 어떻게 생각하시는 지 궁금합니다...
저는 재사용성, 가독성을 위해 컴포넌트를 최대한 많이 쪼개는 것을 지향하는 편이라, 컴포넌트의 개수에 대해서는 크게 문제가 없다고 생각합니다.

다만, 창욱님의 경우에는 컴포넌트 파일들의 depth 가 너무 깊고, 컴포넌트 이름만 보고는 이 컴포넌트가 어떤 역할을 하는지 잘 모르겠는 부분들이 있어서 이해하기 어려웠던 부분들이 있는 것 같습니다.

하지만, 이런 건 프로젝트를 많이 경험하다보면 창욱님께서 언제 어떤 이름으로 컴포넌트를 분리하면 좋을 지 노하우가 쌓이게 되는 휴리스틱한 부분이라, 프로젝트를 계속 진행하시다보면 나아질 것 같다고 생각합니다.

### Q. 최대한 컴포넌트를 직접 렌더링 하지 않고, 상태만 변경하도록 코드를 작성했습니다. 혹시 그렇지 않은 부분이 있다면 가감없이 말씀해주시면 감사하겠습니다.

사실 저는 컴포넌트 패턴의 의의 중 하나가, "rendering 은 state 변경에 종속되어 있다"라는 통일된 규칙에 있다고 생각합니다. 컴포넌트를 랜더링 하기 위해서는 무조건 상태를 변경해야 한다는 규칙을 잘 지키려고 노력하신 부분에 대해서는 칭찬드리고 싶습니다! 하지만, 프로그래스바처럼 어떤 컴포넌트에서 다른 컴포넌트를 querySelector 로 구해서 상태를 변경하는 부분은 수정이 필요할 것 같습니다.

### Q. DocumentList와 Document가 서로를 호출하는 방식으로 구현되어있습니다. 어떤 부작용이 있는지는 잘 모르는 상태인데 문제점이 있다면 말씀 부탁드립니다.

이 부분은 DocumentPage/index.js 에 비슷한 관련 코멘트를 남겨놓았습니다. https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7#discussion_r1257724541

### Q. 성능 개선을 어떻게 해야할 지 잘 모르겠습니다... 비효율적이라고 생각되는 부분이 있다면 말씀 부탁드립니다! ㅠㅠ

저도 이번 프로젝트에서는 성능 개선을 생각할 겨를이 없엇어서... 이 부분에 대해서는 의견을 내기 조심스럽네요...(제가 시도해보지 않은 부분이라...)

### Q. 이런 기능이 있으면 좋을것 같다고 생각하시는 것들도 말씀해주시면 감사하겠습니다.

실제 Notion 처럼 document 페이지 상단에 현재 페이지의 경로를 보여주는 UI 가 있다면 좋을 것 같습니다.
ex) root1/child1/current

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/index.js:
```diff
@@ -0,0 +1,66 @@
+      new Document({ $target: $documentList, initialState: doc }).setState(doc);
+    });
+  };
+}
```
state 값 변화로 구현하신게 인상 깊었어요!
단 index.js 보다는 DocumentList.js 가, style.css보다는 DocumentList.css 가 더 좋을 것 같아요...
파일명에 차별화를 둠으로써 다른 사람들이 볼 때 좀 더 바로바로 이해할 수 있고, 찾기도 쉬울 것 같아요 :)
추가로 delete 구현하실 때 res로 parent를 받아 부모 문서로 이동하는게 더 사용자 측면에서 좋아보이네요 :)
ux 측면에서도 잘 구현하신 것 같습니다!

wukdddang replied:
파일 구조를 시범적으로 index.js로 해봤는데 (webpack에서 파일명이 index.js면 알아서 폴더 명으로 매핑해줌) 가독성이 떨어지네요 ㅠㅠ 많은 사람들이 이렇게 작성하지 않는 이유가 있는 것 같습니다.... 디버그 할때도 힘들구요..
파일명은 지선님 리뷰처럼 바꾸는 게 좋을 것 같아요. 그냥 우직하게 밀고 나갔는데 리팩토링 할 게 많은 것 같네요 ㅋㅋㅋㅋㅋㅋㅋ 🥲🥲 리뷰 감사합니다!! ㅎㅎ ☺️

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/index.js:
```diff
@@ -0,0 +1,49 @@
+
+    $document.appendChild($documentContent);
+  };
+}
```
isOpen 만드셨는데 좋은 것 같아요! 
궁금한 부분이 있는데 button 태그 대신 span 태그를 사용하신 이유가 있을까요?

wukdddang replied:
이유는 없었습니다... 구현하고 보니 버튼으로 되어있었네요 🥲 스터디때 수정하도록 하겠읍니다 ㅠㅠ 
리뷰 감사합니다!! ㅎㅎ ☺️

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/style.css:
```diff
@@ -0,0 +1,65 @@
+  display: flex;
+  justify-content: center;
+  align-items: center;
+}
```
컴포넌트 별로 css 파일을 만들어 관리한 게 제 개인적으로는 좋은 것 같습니다!
이 부분은 하나의 style.css 파일로 다 관리하기도 하고 컴포넌트 별로 관리하기도 한다고 개인차가 심하다고 들었는데 저는 개인적으로 분리해서 개별적으로 관리하는 것이 좀 더 보기 편한 거 같아요 ㅎㅎ
다만 style.css 이름만 Document.css 로 바꾸면 좀 더 다른 사람들도 보기 편하지 않을까 합니다!

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/SideBar/index.js:
```diff
@@ -0,0 +1,91 @@
+
+    const { documentId } = target.closest('.DocumentHeader').dataset;
+    setCurrentDocumentId(documentId);
+  };
```
classList 를 굉장히 잘 활용하시고 계신거 같아요 👍

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/SideBar/index.js:
```diff
@@ -0,0 +1,91 @@
+      push(`/documents/${documentId}`);
+    }
+  });
+}
```
이 파일 또한 index.js 보다는 SideBar.js 로 바꾸시면 어떨까요? ㅎㅎ

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/SideBar/style.css:
```diff
@@ -0,0 +1,15 @@
+  border-radius: 1rem;
+  border: 1px solid #ccc;
+  box-shadow: 2px 4px 4px #ccc;
+}
```
이름을 정말 깔끔하고 직관적으로 지으셔서 이해하기 쉬운 것 같아요.
이런 점 본받고 싶네요 ㅎㅎ
style.css 보다는 SideBar.css 를 조심스럽게 추천해봅니다~

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/SideBarHeader/index.js:
```diff
@@ -0,0 +1,28 @@
+      document.querySelector('.SideBarOpenButton').classList.add('ButtonHide');
+    }
+  });
+}
```
코드가 깔끔해서 보기 좋았어요! 이해하기도 쉽고요 ㅎㅎ
왜 SideBarHeader 로 지었을까 했는데 코드를 보니 바로 이해되네요 ㅎㅎ
이 부분도 index.js와 style.css 보다는 SideBarHeader.js와 SideBarHeader.css 추천드립니당~

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/index.js:
```diff
@@ -0,0 +1,40 @@
+  };
+
+  $target.appendChild($directoryPage);
+}
```
왜 DirectoryPage를 만드셨을까 했는데 코드를 보니 이해가 되네요. 
찬욱님은 컴포넌트를 너무 많이 만드신 것 같아 고민이라고 하셨지만, 저는 오히려 많이 만들면서도 체계적으로 구성하셔서 직관적이고 이해하기가 편했어요. 
이 부분은 다른 사람들과 협업할 때 다들 그렇게 느끼지 않으실까 합니다! 
다만 다른 분들 말씀대로 depth가 좀 있는 편이라... 이 부분은 개선할 수 있다면 좋을 것 같아요! 
(여기까지 오는데 파일 경로가 많이 줄어드는게 보여서 저도 모르게 와 이제 지상이 보인다... 했네요^^;;;;)

Lim-JiSeon commented on src/components/ContentWrapper/DirectoryPage/style.css:
```diff
@@ -0,0 +1,4 @@
+.DirectoryPage {
+  display: flex;
+  flex-direction: column;
+}
```
이 부분 역시 index.js 와 style.css 보다는 DirectoryPage.js와 DirectoryPage.css 를 추천드립니다.
파일명 같은 경우 사람마다 다르고, 취향 성향이 있어서 커맨트를 달기 조심스럽지만, 좀 더 이해하기 쉽고 협업하기 좋은 코드가 될 것 같아 조심스럽게 추천해봅니다~!

Lim-JiSeon commented on src/components/ContentWrapper/DocumentPage/TextEditor/index.js:
```diff
@@ -0,0 +1,42 @@
+
+  $titleEditor.addEventListener('keyup', onKeyUp);
+  $textEditor.addEventListener('keyup', onKeyUp);
+}
```
TextEditor라고 구체적으로 하신 부분 좋은 것 같아요! 저는 단순하게 Editor 한거 같은데 이런 부분 본받고 싶네요 ㅎㅎ
index.js -> TextEditor.js
style.css -> TextEditor.css
이런 식으로 수정하면 찬욱님께서 많이 고민하시면서 지은 컴포넌트 이름을 더 잘 활용할 수 있을 것 같아요 💯

Lim-JiSeon commented on src/components/ContentWrapper/index.js:
```diff
@@ -0,0 +1,28 @@
+    directoryPage.setState();
+    documentPage.setState(this.state);
+  };
+}
```
구체적으로 직관적으로 이름을 지어주시는 점은 너무 좋은 것 같아요!
다만 wrapper 단어 같은 경우 저도 주로 css적인 부분에서만 classname이나 id로만 봤어요. 
컴포넌트명에는 적합하지 않은 것 같아서 다른 이름으로 바꿔보시는 건 어떨까요?

Lim-JiSeon commented on src/components/HeaderWrapper/ProgressBar/style.css:
```diff
@@ -0,0 +1,11 @@
+  background-color: #6ce4b9;
+  z-index: 100;
+  transition: all 0.5s ease-in-out;
+}
```
딜레이 시간을 시각적으로 나타내는 부분이 인상깊었어요! 유저 입장에서도 저장되는 것을 시각적으로 볼 수 있어서 좋은 것 같습니다~!

wukdddang replied:
감사합니다 ㅎㅎ 근데 아직 문제가 많은 것 같아요...🥲

Lim-JiSeon reviewed (commented):
이번 과제 개인 프로젝트 정말 고생하셨습니다!
창욱님께서 많이 고민하시고 고민하신 부분들이 보이는 것 같아요 :)
다른 사람들이 보기에 쉽고, 이해하기 편한 코드인 만큼 창욱님께서 많은 시간과 노력을 들이셨다는 것이 보입니다. 
처음과 비교해 빠른 속도로 발전하고 계신 것같아 저도 본받아야겠다는 생각이 드네요 ^^;;;
다른 분들께서 이미 좋은 말씀 많이 해주셔서 저는 index.js와 style.css 파일명만 수정하고 depth에 대해서 고민해 개선하시면 지금도 좋지만 더 좋아질 거같다란 말 드리고 싶습니다~

hayamaster commented on src/App.js:
```diff
@@ -0,0 +1,30 @@
+    } else if (pathname.indexOf('/documents/') === 0) {
+      const [, , id, status] = pathname.split('/');
+      const res = await request(`/documents/${id}`);
+      setCurrentDocumentId(id);
```
저도 코드를 봤을 땐, 이렇게 현재 Id를 저장하는 전역 변수를 설정할 수도 있구나! 했는데,
혜진님의 피드백을 보니, 이렇게 구현하신 이유가 궁금해졌습니다!

wukdddang replied:
혜진님 리뷰의 밑에 달았는데, 수정하고 나니 코멘트가 Files Changed에는 안보이네요 😭
[제 답변](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7#discussion_r1264488754) 에 내용을 적어두었습니다! 🙇🏻‍♂️🙇🏻‍♂️

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/index.js:
```diff
@@ -0,0 +1,49 @@
+  };
+
+  this.render = () => {
+    const { id, title, documents, isOpen = false } = this.state;
```
isOpen 네이밍 너무 좋은 것 같아요! 
이렇게 boolean값을 갖는 변수는 is~ 와 같은 be동사로 시작하면 의미가 명확해지는 것 같습니다 ㅎㅎ

wukdddang replied:
네이밍 칭찬을 받았다.... 처음 하는 경험이군요 ㅎㅎ 따뜻한 리뷰 감사합니다 ㅎㅎ ☺️☺️

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/index.js:
```diff
@@ -0,0 +1,49 @@
+
+    const $documentContent = document.createElement('span');
+    $documentContent.className = 'DocumentContent';
+    $documentContent.style.display = isOpen ? 'block' : 'none';
```
`li`안에 현재 문서를 담는 `div`와 자식 문서를 담는 `span`을 넣고 `span`에 style을 입혀 주셨군요! 
저는 계속 div나 ul에 style을 입혀주었는데, 오류가 발생해서 결국 아코디언을 구현하지 못했습니다 ㅠㅠ
창욱님께 영감받아 다시 리팩토링 도전해보겠습니다 ㅎㅎㅎㅎ 
좋은 소스 감사합니다ㅎㅎ 🙏

wukdddang replied:
크... 좋은 소스가 될 수 있다면 영광입니다... ㅎㅎ 리팩토링 같이 화이팅입니다! ㅎㅎ ☺️

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/index.js:
```diff
@@ -0,0 +1,49 @@
+      $documentContent.innerText = '하위 문서가 없습니다. 🤔';
+    }
+
+    const displayedTitle = title.length > 20 ? `${title.substring(0, 20)}...` : title;
```
이런 섬세함까지.... 창욱님 센스 💯

wukdddang replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 그러나 멘토님 말씀대로 상수 처리를 해줬어야 했는데... 아쉽습니다... 😭

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/index.js:
```diff
@@ -0,0 +1,66 @@
+    this.render();
+  };
+
+  const toggleDocument = (document, documentId) => {
+    const { id, documents } = document;
+    if (id === Number(documentId)) {
+      return { ...document, isOpen: !document.isOpen };
+    }
+
+    if (Array.isArray(documents) && documents.length > 0) {
+      const updateDocuments = documents.map((childDocument) => toggleDocument(childDocument, documentId));
+      return { ...document, documents: updateDocuments };
+    }
+    return document;
+  };
```
와 재귀함수 진짜 깔끔하게 짜신 것 같습니다!
너무 잘 읽힙니다!

wukdddang replied:
흐흐... 칭찬해주셔서 넘 감사합니다 ㅎㅎㅎ ☺️☺️

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/index.js:
```diff
@@ -0,0 +1,66 @@
+  };
+
+  this.onAdd = async (documentId) => {
+    const res = await request('/documents', {
+      method: 'POST',
+      body: JSON.stringify({ title: '새로운 문서', parent: documentId ? Number(documentId) : null }),
+    });
```
지금도 충분히 깔끔하고 잘하셨지만, "POST", "GET", "PUT", "DELETE" api 처리를 따로 모아두는 컴포넌트를 만들어서 사용해도 좋을 것 같습니다!
혜진님 예시가 가장 적절한 것 같아 [첨부](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/36/files#diff-1952ef38ecd8505652ac151cac71f96e74b799a2e23344fa5bf2e74f3a813ae3)해보겠습니다!

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBar/index.js:
```diff
@@ -0,0 +1,91 @@
+  // 사이드바 닫기
+  this.hide = () => {
+    $sideBar.style.display = 'none';
+  };
```
창욱님 이렇게 함수를 자유자재로 구현하시다니....
너무 존경합니다.✨✨
아이디어 카피좀 하겠습니다ㅎ

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBar/DocumentList/Document/index.js:
```diff
@@ -0,0 +1,49 @@
+    `;
+    if (currentDocumentId && id === Number(currentDocumentId)) {
+      $document.querySelector('.DocumentTitle').classList.add('clicked');
+    }
```
[@components/ContentWrapper/DirectoryPage/SideBar](https://github.com/prgrms-fe-devcourse/FEDC4-5_Project_Notion_VanillaJS/pull/7/files#diff-bc5a68898c0e666b1314ec597ca223c65c06f591a1ad1e5ba7d5d6caded2a82e)의 49번째 줄에서 `target.classList.add('clicked');` 을 해주고 있습니다!
따라서 중복되는 코드라 생각됩니다만, 혹시 다른 의도가 있을까요??
만약 없으시다면 삭제하셔도 무방할 것 같습니다!

wukdddang replied:
이 코드가 없으면 현재 문서 표시가 사라지더라구요... 🥲
안쓰고 싶었는데 도저히 방법이 생각이 안나더라구요 ㅠㅠ Document랑 DocumentList를 재귀적으로 호출해서 그런건가... 싶기도 합니다.. 😭😭

hayamaster commented on src/components/ContentWrapper/DirectoryPage/SideBarHeader/index.js:
```diff
@@ -0,0 +1,28 @@
+      ⏎
+    </div>
+    <div class="SideBarButton SideBarOpenButton ButtonHide">
+      ☰
```
이미지 파일이 아니라 신기한 아이콘 사용 너무 참신하고 좋습니다 ㅎㅎㅎ

hayamaster commented on src/components/ContentWrapper/DocumentPage/index.js:
```diff
@@ -0,0 +1,61 @@
+          textEditor.setState({ status: 'saved', ...createdDocument });
+          updateState({ status: 'saved', ...createdDocument });
+
+          await new Promise((resolve) => setTimeout(resolve, 2000));
```
이 부분은 혹시 무슨 역할을 하는 것일까요?? 
이 코드가 없어도 잘 동작하는 것 같아 질문드립니다!

wukdddang replied:
최대한 기능 구현에 집중한 코드였습니다... 헤더부분에 저장 UI를 표현하려다 보니, 비동기 처리를 구현해줘야 해서 억지로 우겨넣은 코드인데 적을때도 아 이건 좀 이상한데... 라는 생각으로 작성했습니다. ㅠㅠ 😭 리팩토링할 때 다시 뜯어봐야 할 것 같습니다!

hayamaster commented on src/components/ContentWrapper/DocumentPage/index.js:
```diff
@@ -0,0 +1,61 @@
+  });
+
+  this.render = () => {
+    const { documentId, status } = this.state;
+
+    if (!documentId && !status) {
+      $documentPage.style.visibility = 'hidden';
+    } else {
+      $documentPage.style.visibility = 'visible';
+    }
```
이부분은 없어도 무방할 것 같다고 조심스럽게 생각해봅니다.
혹시 의도된 점이 있으면 알려주시면 감사하겠습니다!!!

wukdddang replied:
이 부분은 맨 처음 페이지로 갔을 때 DocumentPage를 보이지 않게 하려고 hidden, visible 속성을 준 코드입니다. 혹시 이렇게 구현하지 않고 더 좋은 방법으로 할 수 있을까요?? 🙇🏻‍♂️🙇🏻‍♂️

hayamaster commented on src/components/HeaderWrapper/ProgressBar/index.js:
```diff
@@ -0,0 +1,8 @@
+import './style.css';
+
+export default function ProgressBar({ $target }) {
+  const $progressBar = document.createElement('div');
+  $progressBar.className = 'ProgressBar';
+
+  $target.appendChild($progressBar);
+}
```
혹시 ProgressBar가 동작을 잘 하는 것일까요? 
제가 렌더링 해보았을 땐, 확인이 안되어서 질문드립니다!

wukdddang replied:
넵 기능은 잘 동작되는 것 같습니다! 
ProgressBar 기능 구현할 때 로직을 복잡하게 해서 수정이 필요할 것 같습니다... 😭😭

hayamaster reviewed (commented):
창욱님! 이번엔 진짜 코드를 읽으면서 감탄을 금치 못했습니다.
진짜 너무너무너무 잘하셨어요. 코드 너무 잘읽히고, 컴포넌트화까지 완벽한 것 같습니다.
진짜 너무 빨리 성장하시는 것 같아 무서울 정도 입니다....😭
코드 한줄 한줄 깊이 생각하시고 정성을 들였다는 것이 너무 많이 느껴졌어요. 
정말 너무 고생하셨고, 좋은 코드 읽을 수 있는 기회를 주셔서 감사합니다!

너무 잘하셔서 의미있는 피드백을 드렸는지 모르겠습니다... 대부분 칭찬이나 질문을 한 것 같습니다ㅠㅠ 죄송합니다

- 처음에 HeaderWrapper 폴더 안에서는 좀 과하게 컴포넌트를 나눈 것은 아닌가 생각했습니다. 하지만 오히려 나눠주셔서 더욱 가독성있는 것 같습니다!! 좋아요! 다른 분 말씀처럼 depth는 조금 들어가는 느낌이 있지만 잘읽히니 좋습니다 ㅎㅎ

</br>

\+ 추가로 조금만 손봐주시면 정말 완벽한 프로젝트가 될 것 같습니다!
1. 뒤로가기를 눌렀을 때, `DocumentPage`가 다시 렌더링 되지 않습니다. 이 부분을 수정해주시면 좋을 것 같습니다!
2. Root document를 생성하는 기능은 빠져있는 것 같습니다! 함께 추가해주시면 더욱 좋은 UX를 가질 수 있을 것 같습니다ㅎㅎ

wukdddang committed:
Refactor: /documents url 처리 제거
- 과제 요구사항에 명세되지 않은 기능
- root로 이동하도록 처리

wukdddang committed:
Refactor: eslint 설정 변경
- no-new는 headerWrapper에서만 비활성화 하므로, 전역에서 필요없는 설정
- no-unresolved는 없는 설정

wukdddang committed:
Refactor: EOL 처리
-prettier에 잡히지 않는 .prettierrc.js, .gitignore파일 수정

wukdddang committed:
Refactor: EOF 오류 처리
- .editorconfig를 사용하여 EOF 오류 처리
- .prettierrc.js에 해당 설정 추가

wukdddang committed:
Refactor: App 생성 코드 수정
- initApp 함수로 eslint의 no-new 설정 유지

wukdddang committed:
Refactor: App 컴포넌트 수정
- 불필요한 setCurrentDocumentId 제거
- 이유: window.location으로 현재 문서 id를 알 수 있음

wukdddang committed:
Refactor: Document 컴포넌트 수정
- DocumentList와 같은 Level에 있도록 수정

wukdddang committed:
Refactor: Header 컴포넌트 수정
- style.css에 있던 font 설정 제거

wukdddang committed:
Refactor: 토글 버튼 클릭 시 push 하는 기능 제거

wukdddang committed:
Fix: .editorconfig 파일 수정 (EOF)

wukdddang committed:
Refactor: status 변수명 isDocumentSaved로 변경

wukdddang committed:
Refactor: DocumentList 수정
- 토글 클릭 시 push되던 현상 수정

wukdddang committed:
Refactor: 사이드바 여닫는 기능 제거
- SideBar 컴포넌트가 open/close인 상태를 가질 수 있도록 전체 애플리케이션 구조 변경이 필요함
- 이번주 리팩토링 기간에 애플리케이션 재구성 에정

wukdddang committed:
Refactor: utils/storage/index.js 함수명 변경
- getOpenDocumentIds -> 좀 더 시멘틱한 이름으로 변경 getToggledDocumentsIds
-getDocuments -> getToggledDocuments
- set... 관련 함수도 동일하게 처리

wukdddang committed:
Refactor: DocumentPage 컴포넌트 수정
- status: string이 아닌 isDocumentSaved: boolean으로 state 변경

wukdddang committed:
Refactor: async / await로 받은 데이터 결과 체크 코드 추가

wukdddang committed:
Refactor: async / await로 받은 데이터 결과 체크 코드 추가 (2)

wukdddang committed:
Refactor: popstate 기능 추가
